### PR TITLE
MB-33556: Call the OnFlushCompleted handler unconditionally

### PR DESCRIPTION
**Pull request by BenHuddleston**
In the case of a shutdown or other reasons in which a flush may
not complete, OnFlushCompleted is not called because the Status is
not OK. Instead, pass the status through to the event listener via the
FlushJobInfo and allow the listener to deal with any failure as they
see fit.

### CODE REVIEW

File: db/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -511,7 +513,8 @@ Status DBImpl::AtomicFlushMemTablesToOutputFiles(
 
 void DBImpl::NotifyOnFlushBegin(ColumnFamilyData* cfd, FileMetaData* file_meta,
                                 const MutableCFOptions& mutable_cf_options,
-                                int job_id, TableProperties prop) {
+                                const Status& st, int job_id,
```
Comment (daverigby) [inline db/db_impl_compaction_flush.cc]: I don't think it makes much sense to pass Status into `NotifyOnFlushBegin` given it's always going to be default-constructed (I assume ok?). 

Simpler to revert the change to the method, and just default-construct it at line 539.

**Review by daverigby — CHANGES_REQUESTED** [review]
You also need to update all implementations of `OnFlushCompleted` to do nothing if status is not ok - so their behaviour is the same as previously.

File: db/db_impl_compaction_flush.cc (hunk 1 above)
Comment (BenHuddleston) [reply db/db_impl_compaction_flush.cc]: Defaults to ok. I just copied the OnCompaction**** API which passes a status into OnCompactionBegin, but I did some digging and they have some pre-condition that you have to meet before OnCompactionBegin is called so the status isn't necessarily OK.  Reverted the NotifyOnFlushBegin signature change, manually set the status in the function to OK for clarity.

**Comment by BenHuddleston** [comment]
> You also need to update all implementations of `OnFlushCompleted` to do nothing if status is not ok - so their behaviour is the same as previously.

Done

**Review by daverigby — APPROVED** [review]
Suggest you also raise an issue on the upstream and propose this as a pull request.
